go-common.mk

```diff
@@ -181,25 +169,23 @@ $(GOTESTCOVERHTML): $(GOTESTCOVERRAW)
 	esac
 
 # Go executables
-$(_GO_ROOT_BUILD_TARGET): $(GOSRC) go.mod
+$(_GO_ROOT_BUILD_TARGET): $(GOSRC) generate dependencies
 	@-mkdir build 2> /dev/null
-ifdef GORUNGENERATE
-	$(GO) generate ./...
-endif # GORUNGENERATE
-ifdef GORUNGET
-	$(GO) get ./...
-endif # GORUNGET
 	$(GOBUILDENV) $(GO) build $(GOBUILDFLAGS) -o $@ .
 
-$(_GO_BUILD_TARGETS): $(GOSRC) go.mod
+$(_GO_BUILD_TARGETS): $(GOSRC) generate dependencies
 	@-mkdir $(BUILDDIR) 2> /dev/null
+	$(GOBUILDENV) $(GO) build $(GOBUILDFLAGS) -o $@ $(GOCMDDIR)/$(notdir $@)
+
+dependencies: go.mod
+ifdef GORUNGET
+	$(GO) get -t ./...
+endif # GORUNGET
```

Comment:
We've tended to mostly name the targets after the tools they invoke (lint is perhaps an exception, although I would argue it is substantively meaningful), so should this target be 'get' instead of 'dependencies'?

This also changes behaviour a little in that test dependencies are always fetched whether we are running tests or not.  I think that is ok, but wanted see if that was of concern for anyone.

Perhaps the biggest question is that unlike the other top level targets, we're not following the same pre-/standard-/post- pattern that enables those steps to be overridden.  We've used those sparsely in our code base, but there have been cases where they've been useful.  Should we follow that pattern for these targets as well, or are we asserting that there are never useful pre or post steps for them?

Reply:
I have a question about the pre, standard, post thing. It seems like a bunch of those don't have anything in the pre or post targets, are those there to maintain the pattern? If so, I could see wanting to do that for this as well

Reply:
So re: the `-t` flag, I did a little experimenting.

First I tried removing the flag (and clearing my modcache) and found that the `go test` command actually installs the missing test dependencies on its own - we don't need to explicitly use `go get -t` beforehand.

Then I got curious and went a step further, and deleted the `get` recipe and all references to `go get` in the Makefile entirely. I went through the main recipes (test, lint, build) and repeatedly cleared my modcache before trying each one. I found is that Go will consistently just install dependencies as needed, and I wasn't able to contrive any scenario in which it was necessary to call `go get` directly. The situation doesn't appear to be any different in circle, although fully testing that out by clearing the modcache would be destructive, since I think we share that cache between builds.

@glewis-ANet do you know of any other reasons to have explicit `go get` commands, or have concerns about removing them?

Reply:
> So re: the `-t` flag, I did a little experimenting.
> 
> First I tried removing the flag (and clearing my modcache) and found that the `go test` command actually installs the missing test dependencies on its own - we don't need to explicitly use `go get -t` beforehand.
> 
> Then I got curious and went a step further, and deleted the `get` recipe and all references to `go get` in the Makefile entirely. I went through the main recipes (test, lint, build) and repeatedly cleared my modcache before trying each one. I found is that Go will consistently just install dependencies as needed, and I wasn't able to contrive any scenario in which it was necessary to call `go get` directly. The situation doesn't appear to be any different in circle, although fully testing that out by clearing the modcache would be destructive, since I think we share that cache between builds.
> 
> @glewis-ANet do you know of any other reasons to have explicit `go get` commands, or have concerns about removing them?

I have no objection to removing the explicit `go get` commands.  I don't see any overrides of the variables associated with that, so if we're confident that `go build` and `go test` will do the right thing without that then I'm happy with all that being gone.